process/src/services/service.ports.ts

```diff
@@ -1,295 +1,379 @@
-import { ControllerSerialPort, IOptions, IIOState } from '../controllers/controller.serialport';
-import * as SerialPort from 'serialport';
-import Logger from '../env/env.logger';
-import PluginIPCService from 'chipmunk.plugin.ipc';
-import { IPCMessages } from 'chipmunk.plugin.ipc';
-import { ERenderEvents } from '../consts/events';
+import {
+  ControllerSerialPort,
+  IOptions,
+  IIOState,
+} from "../controllers/controller.serialport";
+import * as SerialPort from "serialport";
+import Logger from "../env/env.logger";
+import PluginIPCService from "chipmunk.plugin.ipc";
+import { IPCMessages } from "chipmunk.plugin.ipc";
+import { ERenderEvents } from "../consts/events";
 
 export interface IListeners {
-    onData: (chunk: Buffer) => void;
-    onError: (error: Error) => void;
-    onDisconnect: () => void;
+  onData: (chunk: Buffer) => void;
+  onError: (error: Error) => void;
+  onDisconnect: () => void;
 }
 
 export interface IPortInfo {
-    path: string;
-    manufacturer?: string;
-    serialNumber?: string;
-    pnpId?: string;
-    locationId?: string;
-    productId?: string;
-    vendorId?: string;
+  path: string;
+  manufacturer?: string;
+  serialNumber?: string;
+  pnpId?: string;
+  locationId?: string;
+  productId?: string;
+  vendorId?: string;
 }
 
 export interface IPortState {
-    ioState: IIOState;
-    connections: number;
+  ioState: IIOState;
+  connections: number;
 }
 
 class ServicePorts {
+  private _controllers: Map<string, ControllerSerialPort> = new Map();
+  private _listeners: Map<string, Map<string, IListeners>> = new Map();
+  private _logger: Logger = new Logger("ControllerSerialPortManager");
+  private _token: string | undefined;
+  private _connectedPortsState: {
+    timer: any;
+    attempts: number;
+    state: { [key: string]: IPortState };
+  } = {
+    timer: -1,
+    attempts: 0,
+    state: {},
+  };
 
-    private _controllers: Map<string, ControllerSerialPort> = new Map();
-    private _listeners: Map<string, Map<string, IListeners>> = new Map();
-    private _logger: Logger = new Logger('ControllerSerialPortManager');
-    private _token: string | undefined;
-    private _connectedPortsState: {
-        timer: any,
-        attempts: number,
-        state: { [key: string]: IPortState }
-    } = {
-        timer: -1,
-        attempts: 0,
-        state: {}
-    };
+  constructor() {}
 
-    constructor() {
+  public destroy(): Promise<void> {
+    return new Promise((resolve) => {
+      this._listeners.clear();
+      Promise.all(
+        Array.from(this._controllers.keys()).map((port: string) => {
+          return this._close(port);
+        })
+      ).then(() => {
+        this._controllers.clear();
+        resolve();
+      });
```

Comment:
I think here are a couple of potential issues:
1) `catch` is missed. If some `this._close(port)` would be rejected; `Promise.all` will drop `catch`
2) If some `this._close(port)` would be rejected; `Promise.all` will drop `catch` and not all ports would be closed. Let's say we have opened: A, B, C and D; we are trying to close all of them like it:
A - closed
B - fail (reject)
C - pending
D - pending
As soon as B was failed, C and D might be still opened. 
Maybe better to use here `Promise.allSettled()` instead `Promise.all()`